File: KIWI/Service/DataCompareService.vb
Imports IDAUtil

Public Class DataCompareService

    Public Function getSalesDocumentToChangeQuantity(zvo4iList As List(Of ZV04IProperty), shipToConversionList As List(Of ConversionShiptTo), materialConversionList As List(Of ConversionMaterial)) As List(Of QtyConversionOrderProperty)
        'TODO: move to function
        Dim zv04IChangeList = From zv04Item In zvo4iList Where zv04Item.rejReason = ""
                              Join shipConversion In shipToConversionList On zv04Item.shipTo Equals shipConversion.shipTo
                              Join materialConversion In materialConversionList On zv04Item.material Equals materialConversion.material
                              Select zv04Item.order, zv04Item.shipToName, zv04Item.soldToName, zv04Item.soldTo, zv04Item.item, zv04Item.shipTo, zv04Item.material, zv04Item.orderQty, materialConversion.conversionIndex

        'TODO: refactor
        Dim baseSalesDocumentQuery =
            From q In zv04IChangeList
            Group q By line = New With {Key q.order, q.shipTo, q.soldTo, q.shipToName, q.soldToName}
                                    Into Group
            Select New QtyConversionOrderProperty With {
            .orderNumber = line.order,
            .shipTo = line.shipTo,
            .soldTo = line.soldTo,
            .shipToName = line.shipToName,
            .soldToName = line.soldToName,
            .documentLineList = Group.Aggregate(New List(Of DocumentLine), Function(x, q)
                                                                               x.Add(New DocumentLine() With {.item = q.item, .material = q.material, .quantity = q.orderQty})
                                                                               Return x
                                                                           End Function),
            .documentLineChangeList = Group.Aggregate(New List(Of DocumentLine), Function(x, q)
                                                                                     x.Add(New DocumentLine() With {.item = q.item, .material = q.material, .quantity = q.orderQty * q.conversionIndex})
                                                                                     Return x
                                                                                 End Function)
            }

        Return baseSalesDocumentQuery.ToList()
    End Function

    Public Function removeAlreadyConvertedLines(logConversionList As List(Of ConversionLog), ByRef salesDocumetList As List(Of QtyConversionOrderProperty)) As List(Of QtyConversionOrderProperty)
        For i = salesDocumetList.Count - 1 To 0 Step -1
            For Each logLine In logConversionList
                'delete line if was already converted
                If salesDocumetList(i).orderNumber = logLine.orderNumber AndAlso isSuccessfullyConvertedLineAlreadyInLog(salesDocumetList(i), logLine) Then
                    salesDocumetList(i).removeLineByItem(logLine.item)
                    'delete document if all items in the document were changed priviously
                    If salesDocumetList(i).documentLineChangeList.Count = 0 Then
                        salesDocumetList.RemoveAt(i)
                    End If
                End If
            Next
        Next
        Return salesDocumetList
    End Function

    Private Function isSuccessfullyConvertedLineAlreadyInLog(document As QtyConversionOrderProperty, logLine As ConversionLog) As Boolean
        For i = document.documentLineChangeList.Count - 1 To 0 Step -1
            If document.documentLineChangeList(i).item = logLine.item AndAlso logLine.isConverted AndAlso logLine.isSaved Then
                Return True
            End If
        Next
        Return False
    End Function

    <Obsolete>
    Public Function getNotPreviouslyChangedDocuments(salesDocumentList As List(Of QtyConversionOrderProperty), logConversionList As List(Of ConversionLog)) As List(Of QtyConversionOrderProperty)
        For i = salesDocumentList.Count - 1 To 0 Step -1
            For j = 0 To logConversionList.Count - 1
                If logConversionList(j).isConverted AndAlso salesDocumentList(i).orderNumber = logConversionList(j).orderNumber Then
                    salesDocumentList.RemoveAt(i)
                End If
            Next
        Next
        Return salesDocumentList
    End Function

End Class

File: KIWI/Service/DataCompareService.cs
using IDAUtil;
using IDAUtil.Model.Properties.TcodeProperty.ZV04Obj;
using Microsoft.VisualBasic.CompilerServices;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace KIWI {
    public class DataCompareService {
        public List<QtyConversionOrderProperty> getSalesDocumentToChangeQuantity(List<ZV04IProperty> zvo4iList, List<ConversionShiptTo> shipToConversionList, List<ConversionMaterial> materialConversionList) {
            // TODO: move to function
            var zv04IChangeList = from zv04Item in zvo4iList
                                  where string.IsNullOrEmpty(zv04Item.rejReason)
                                  join shipConversion in shipToConversionList on zv04Item.shipTo equals shipConversion.shipTo
                                  join materialConversion in materialConversionList on zv04Item.material equals materialConversion.material
                                  select new { zv04Item.order, zv04Item.shipToName, zv04Item.soldToName, zv04Item.soldTo, zv04Item.item, zv04Item.shipTo, zv04Item.material, zv04Item.orderQty, materialConversion.conversionIndex };

            // TODO: refactor
            var baseSalesDocumentQuery = from q in zv04IChangeList
                                         group q by new { q.order, q.shipTo, q.soldTo, q.shipToName, q.soldToName } into line
                                         select new QtyConversionOrderProperty() {
                                             orderNumber = line.Key.order,
                                             shipTo = line.Key.shipTo,
                                             soldTo = line.Key.soldTo,
                                             shipToName = line.Key.shipToName,
                                             soldToName = line.Key.soldToName,
                                             documentLineList = line.Aggregate(new List<DocumentLine>(), (x, q) => {
                                                 x.Add(new DocumentLine() { item = Conversions.ToInteger(q.item), material = Conversions.ToInteger(q.material), quantity = Conversions.ToInteger(q.orderQty) });
                                                 return x;
                                             }),
                                             documentLineChangeList = line.Aggregate(new List<DocumentLine>(), (x, q) => {
                                                 x.Add(new DocumentLine() { item = Conversions.ToInteger(q.item), material = Conversions.ToInteger(q.material), quantity = (int)(q.orderQty * (double)q.conversionIndex) });
                                                 return x;
                                             })
                                         };
            return baseSalesDocumentQuery.ToList();
        }

        public List<QtyConversionOrderProperty> removeAlreadyConvertedLines(List<ConversionLog> logConversionList, ref List<QtyConversionOrderProperty> salesDocumetList) {
            for (int i = salesDocumetList.Count - 1; i >= 0; i -= 1) {
                foreach (var logLine in logConversionList) {
                    // delete line if was already converted
                    if (salesDocumetList[i].orderNumber == logLine.orderNumber && isSuccessfullyConvertedLineAlreadyInLog(salesDocumetList[i], logLine)) {
                        salesDocumetList[i].removeLineByItem(logLine.item);
                        // delete document if all items in the document were changed previously
                        if (salesDocumetList[i].documentLineChangeList.Count == 0) {
                            salesDocumetList.RemoveAt(i);
                            
                            //exists function if all items were removed out of the list
                            if (salesDocumetList.Count == 0) { return salesDocumetList;}

                            break;
                        }
                    }
                }
            }
            return salesDocumetList;
        }

        private bool isSuccessfullyConvertedLineAlreadyInLog(QtyConversionOrderProperty document, ConversionLog logLine) {
            for (int i = document.documentLineChangeList.Count - 1; i >= 0; i -= 1) {
                if (document.documentLineChangeList[i].item == logLine.item && logLine.isConverted && logLine.isSaved) {
                    return true;
                }
            }

            return false;
        }
    }
}